R/tmb.R

```diff
@@ -81,66 +138,32 @@ h_mmrm_tmb_formula_parts <- function(formula) {
   }
   terms_list <- attr(terms_object, "variables")
   cov_term <- terms_list[[unlist(cov_selected) + 1L]]
-
   # Remove the covariance term to obtain the model formula.
   model_formula <- stats::update(
     formula,
     stats::as.formula(paste(". ~ . -", deparse(cov_term)))
   )
-
-  # Parse the covariance term to get covariance type, visit and subject variables.
-  assert_true(identical(length(cov_term), 2L)) # 2 because `fun (...)`.
-  cov_content <- cov_term[[2L]]
-  if (!identical(length(cov_content), 3L) || !identical(as.character(cov_content[[1L]]), "|")) {
-    stop(
-      "Covariance structure must be of the form `",
-      cov_term[[1]],
-      "(time|(group/)subject)`"
-    )
-  }
-  visit_term <- cov_content[[2L]]
-  if (identical(length(cov_content[[3L]]), 1L)) {
-    subject_term <- cov_content[[3L]]
-    group_var <- NULL
-  } else if (identical(length(cov_content[[3L]]), 3L) && identical(as.character(cov_content[[3L]][[1L]]), "/")) {
-    subject_term <- cov_content[[3L]][[3L]]
-    group_term <- cov_content[[3L]][[2L]]
-    assert_true(identical(length(group_term), 1L))
-    group_var <- deparse(group_term)
-  } else {
-    stop(
-      "Covariance structure with grouping must be of the form `",
-      cov_term[[1]],
-      "(time|group/subject)`"
-    )
-  }
-  assert_true(identical(length(visit_term), 1L))
-  assert_true(identical(length(subject_term), 1L))
-  subject_var <- deparse(subject_term)
-  visit_var <- deparse(visit_term)
+  cov_vars <- h_mmrm_tmb_extract_vars(cov_term)
   full_formula <- stats::update(
     model_formula,
-    stats::as.formula(paste(". ~ . +", subject_var, "+", visit_var))
+    stats::as.formula(paste(". ~ . +", cov_vars$subject_var, "+", cov_vars$visit_var))
   )
-  if (!is.null(group_var)) {
+  if (!is.null(cov_vars$group_var)) {
     full_formula <- stats::update(
       full_formula,
-      stats::as.formula(sprintf(". ~ . + %s", group_var))
+      stats::as.formula(paste(". ~ . +", cov_vars$group_var))
     )
   }
-  ret <- list(
-    formula = formula,
-    model_formula = model_formula,
-    full_formula = full_formula,
-    cov_type = deparse(cov_term[[1L]]),
-    visit_var = visit_var,
-    subject_var = subject_var
-  )
-  if (!is.null(group_var)) {
-    ret <- c(ret, list(group_var = group_var))
-  }
   structure(
-    .Data = ret,
+    .Data = list(
```

Comment:
```suggestion
    list(
```

Reply:
why it is necessary to remove `.Data` argument name?

Reply:
just a minor suggestion - by convention I would say in this case no need to name the first argument, as it is clear that the first one is the object and the other named arguments are the attributes. Note also that it starts with `.` meaning that usually it won't be named explicitly.